deps/apiinfo.go

```diff
@@ -75,20 +79,96 @@ func getFullNodeAPIV1Curio(ctx *cli.Context, ainfoCfg []string, opts ...cliutil.
 		}
 	}
 
-	var v1API api.FullNodeStruct
-	cliutil.FullNodeProxy(fullNodes, &v1API)
+	var v1API api.ChainStruct
+	FullNodeProxy(fullNodes, &v1API)
 
 	v, err := v1API.Version(ctx.Context)
 	if err != nil {
 		return nil, nil, err
 	}
-	if !v.APIVersion.EqMajorMinor(api.FullAPIVersion1) {
-		return nil, nil, xerrors.Errorf("Remote API version didn't match (expected %s, remote %s)", api.FullAPIVersion1, v.APIVersion)
+	if !v.APIVersion.EqMajorMinor(lapi.FullAPIVersion1) {
+		return nil, nil, xerrors.Errorf("Remote API version didn't match (expected %s, remote %s)", lapi.FullAPIVersion1, v.APIVersion)
 	}
 	return &v1API, finalCloser, nil
 }
 
+type contextKey string
+
+// Not thread safe
+func OnSingleNode(ctx context.Context) context.Context {
+	return context.WithValue(ctx, contextKey("retry-node"), new(*int))
+}
+
 type httpHead struct {
 	addr   string
 	header http.Header
 }
+
+var RPCErrors = jsonrpc.NewErrors()
+
+// newChainNodeRPCV1 creates a new http jsonrpc client.
+func newChainNodeRPCV1(ctx context.Context, addr string, requestHeader http.Header, opts ...jsonrpc.Option) (api.Chain, jsonrpc.ClientCloser, error) {
+	var res api.ChainStruct
+	closer, err := jsonrpc.NewMergeClient(ctx, addr, "Filecoin",
+		api.GetInternalStructs(&res), requestHeader, append([]jsonrpc.Option{jsonrpc.WithErrors(RPCErrors)}, opts...)...)
+
+	return &res, closer, err
+}
+
+func FullNodeProxy[T api.Chain](ins []T, outstr *api.ChainStruct) {
```

Comment:
Oh I really like bringing this here, should allow us to eventually port smarts from https://github.com/filecoin-project/lotus/pull/11470 here + add some more Curio-specific ones.

Reply:
👍